central/graphql/resolvers/image_vulnerabilities.go

```diff
@@ -101,3 +212,323 @@ func withImageCveTypeFiltering(q string) string {
+	imageLoader, err := loaders.GetImageLoader(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	q := search.EmptyQuery()
+	q.Pagination = &v1.QueryPagination{
+		Limit:  1,
+		Offset: 0,
+		SortOptions: []*v1.QuerySortOption{
+			{
+				Field:    search.ImageScanTime.String(),
+				Reversed: true,
+			},
+		},
+	}
+
+	images, err := imageLoader.FromQuery(resolver.withImageVulnerabilityScope(ctx), q)
+	if err != nil || len(images) == 0 {
+		return nil, err
+	} else if len(images) > 1 {
+		return nil, errors.New("multiple images matched for last scanned image vulnerability query")
+	}
+
+	return timestamp(images[0].GetScan().GetScanTime())
+}
+
+func (resolver *imageCVEResolver) Vectors() *EmbeddedVulnerabilityVectorsResolver {
+	if val := resolver.data.GetCveBaseInfo().GetCvssV3(); val != nil {
+		return &EmbeddedVulnerabilityVectorsResolver{
+			resolver: &cVSSV3Resolver{resolver.ctx, resolver.root, val},
+		}
+	}
+	if val := resolver.data.GetCveBaseInfo().GetCvssV2(); val != nil {
+		return &EmbeddedVulnerabilityVectorsResolver{
+			resolver: &cVSSV2Resolver{resolver.ctx, resolver.root, val},
+		}
+	}
+	return nil
+}
+
+func (resolver *imageCVEResolver) VulnerabilityState(ctx context.Context) string {
+	if resolver.data.GetSnoozed() {
+		return storage.VulnerabilityState_DEFERRED.String()
+	}
+
+	var imageID string
+	scope, hasScope := scoped.GetScopeAtLevel(resolver.ctx, v1.SearchCategory_IMAGES)
+	if hasScope {
+		imageID = scope.ID
+	}
+
+	if imageID == "" {
+		return ""
+	}
+
+	imageLoader, err := loaders.GetImageLoader(ctx)
+	if err != nil {
+		log.Error(errors.Wrap(err, "getting image loader"))
+		return ""
+	}
+	img, err := imageLoader.FromID(ctx, imageID)
+	if err != nil {
+		log.Error(errors.Wrapf(err, "fetching image with id %s", imageID))
+		return ""
+	}
+
+	states, err := resolver.root.vulnReqQueryMgr.VulnsWithState(resolver.ctx,
+		common.VulnReqScope{
+			Registry: img.GetName().GetRegistry(),
+			Remote:   img.GetName().GetRemote(),
+			Tag:      img.GetName().GetTag(),
+		})
+	if err != nil {
+		log.Error(errors.Wrapf(err, "fetching vuln requests for image %s/%s:%s", img.GetName().GetRegistry(), img.GetName().GetRemote(), img.GetName().GetTag()))
+		return ""
+	}
+	if s, ok := states[resolver.data.GetId()]; ok {
+		return s.String()
+	}
+
+	return storage.VulnerabilityState_OBSERVED.String()
+}
+
+func (resolver *imageCVEResolver) ActiveState(ctx context.Context, args RawQuery) (*activeStateResolver, error) {
+	scopeQuery, err := args.AsV1QueryOrEmpty()
+	if err != nil {
+		return nil, err
+	}
+	deploymentID := getDeploymentScope(scopeQuery, ctx, resolver.ctx)
+	if deploymentID == "" {
+		return nil, nil
+	}
+	// We only support OS level component. The active state is not determined if there is no OS level component associate with this vuln.
+	query := search.NewQueryBuilder().AddExactMatches(search.CVEID, resolver.data.GetId()).AddStrings(search.ComponentSource, storage.SourceType_OS.String()).ProtoQuery()
```

Comment:
We can avoid this query by checking if the operating system field of this image vuln is not set. @connorgorman confirm.

Reply:
No, we still need this query because this checks that the components causing this vuln are from the OS and not languages. Just having the OS will also exist for language vulns because they are coming from the result of that OS

Reply:
@connorgorman This is an important point. There seems to be a big gap that existed in dackbox since `Source Type` field actually did not exist on any top-level protos—storage.ImageComponent and edges. So the queries never would have worked.

Reply:
Found another field `executables` that never existed in the top-level proto.

Reply:
Also, for any given image, the component may come from lang as well as OS. Therefore, seems like the current data model as well as scanner interpretation seems to not fully address the issue.